iroh-bytes/src/baomap.rs

```diff
@@ -0,0 +1,288 @@
+    /// more chunks.
+    fn available_ranges(&self) -> BoxFuture<'_, io::Result<RangeSet2<ChunkNum>>>;
+    /// A future that resolves to a reader that can be used to read the outboard
+    fn outboard(&self) -> BoxFuture<'_, io::Result<D::Outboard>>;
+    /// A future that resolves to a reader that can be used to read the data
+    fn data_reader(&self) -> BoxFuture<'_, io::Result<D::DataReader>>;
+}
+
+/// A generic collection of blobs with precomputed outboards
+pub trait Map: Clone + Send + Sync + 'static {
+    /// The outboard type. This can be an in memory outboard or an outboard that
+    /// retrieves the data asynchronously from a remote database.
+    type Outboard: bao_tree::io::fsm::Outboard;
+    /// The reader type.
+    type DataReader: AsyncSliceReader;
+    /// The entry type. An entry is a cheaply cloneable handle that can be used
+    /// to open readers for both the data and the outboard
+    type Entry: MapEntry<Self>;
+    /// Get an entry for a hash.
+    ///
+    /// This can also be used for a membership test by just checking if there
+    /// is an entry. Creating an entry should be cheap, any expensive ops should
+    /// be deferred to the creation of the actual readers.
+    ///
+    /// It is not guaranteed that the entry is complete. A [PartialMap] would return
+    /// here both complete and partial entries, so that you can share partial entries.
+    fn get(&self, hash: &Hash) -> Option<Self::Entry>;
+}
+
+/// A partial entry
+pub trait PartialMapEntry<D: PartialMap>: MapEntry<D> {
+    /// A future that resolves to an writeable outboard
+    fn outboard_mut(&self) -> BoxFuture<'_, io::Result<D::OutboardMut>>;
+    /// A future that resolves to a writer that can be used to write the data
+    fn data_writer(&self) -> BoxFuture<'_, io::Result<D::DataWriter>>;
+}
+
+/// A mutable bao map
+pub trait PartialMap: Map {
+    /// The outboard type to write data to the partial entry.
+    type OutboardMut: bao_tree::io::fsm::OutboardMut;
+    /// The writer type to write data to the partial entry.
+    type DataWriter: iroh_io::AsyncSliceWriter;
+    /// A partial entry. This is an entry that is writeable and possibly incomplete.
+    ///
+    /// It must also be readable.
+    type PartialEntry: PartialMapEntry<Self>;
+
+    /// Get an existing partial entry, or create a new one.
+    ///
+    /// We need to know the size of the partial entry. This might produce an
+    /// error e.g. if there is not enough space on disk.
+    fn get_or_create_partial(&self, hash: Hash, size: u64) -> io::Result<Self::PartialEntry>;
+
+    /// Get an existing partial entry.
+    fn get_partial(&self, hash: &Hash) -> Option<Self::PartialEntry>;
+
+    /// Upgrade a partial entry to a complete entry.
+    fn insert_complete(&self, entry: Self::PartialEntry) -> BoxFuture<'_, io::Result<()>>;
+}
+
+/// Extension of BaoMap to add misc methods used by the rpc calls.
+pub trait ReadonlyStore: Map {
```

Comment:
nit: maybe `ReadableStore`? 

Also what's the difference in your naming between `Map`s and `Store`s?

Reply:
A map is just a map, a store adds all the stuff you need for a full blown store, such as validation and export.